File: offchain/pkg/roles/roles.go
package roles

// Role constants
const (
	SuperAdmin = "superAdmin"
	NLA        = "nla"
	Admin      = "admin"
	Manager    = "manager"
	Agent      = "agent"
	Notary     = "notary"
	Blocker    = "blocker"
	Paterns    = "paterns"
	Buyer      = "buyer"
)

// RoleHierarchy defines role priority (higher = more powerful)
var RoleHierarchy = map[string]int{
	SuperAdmin: 9,
	NLA:        8,
	Admin:      7,
	Manager:    6,
	Agent:      5,
	Notary:     4,
	Blocker:    3,
	Paterns:    2,
	Buyer:      1,
}

// AllRoles list of all valid roles
var AllRoles = []string{
	SuperAdmin,
	NLA,
	Admin,
	Manager,
	Agent,
	Notary,
	Blocker,
	Paterns,
	Buyer,
}

// GetRoleLevel returns the hierarchy level of a role
func GetRoleLevel(role string) int {
	if level, exists := RoleHierarchy[role]; exists {
		return level
	}
	return 0
}

// CanAssignRole checks if a user with sourceRole can assign targetRole
// Rules:
// - superAdmin and nla can assign any role
// - other users can only assign roles below their level
// - cannot assign a role equal to or higher than their own
func CanAssignRole(sourceRole, targetRole string) bool {
	// superAdmin and nla can assign any role
	if sourceRole == SuperAdmin || sourceRole == NLA {
		return true
	}

	sourceLevel := GetRoleLevel(sourceRole)
	targetLevel := GetRoleLevel(targetRole)

	// Can only assign roles below their level
	return sourceLevel > targetLevel
}

// IsValidRole checks if a role is valid
func IsValidRole(role string) bool {
	_, exists := RoleHierarchy[role]
	return exists
}

// GetHighestRole returns the highest role from a list of roles
func GetHighestRole(userRoles []string) string {
	if len(userRoles) == 0 {
		return ""
	}

	highest := userRoles[0]
	highestLevel := GetRoleLevel(highest)

	for _, role := range userRoles[1:] {
		level := GetRoleLevel(role)
		if level > highestLevel {
			highest = role
			highestLevel = level
		}
	}

	return highest
}

File: offchain/pkg/roles/roles.py
"""
Role definitions and utilities
"""

from typing import List

# Role constants
ADMIN = "admin"
BUYER = "buyer"
SELLER = "seller"
AGENT = "agent"
MODERATOR = "moderator"
SUPER_ADMIN = "super_admin"

# All valid roles
ALL_ROLES = [ADMIN, BUYER, SELLER, AGENT, MODERATOR, SUPER_ADMIN]


def is_valid_role(role: str) -> bool:
    """Check if a role is valid"""
    return role in ALL_ROLES


def validate_roles(roles: List[str]) -> bool:
    """Validate a list of roles"""
    return all(is_valid_role(role) for role in roles)


def has_role(user_roles: List[str], required_role: str) -> bool:
    """Check if user has a specific role"""
    return required_role in user_roles


def has_any_role(user_roles: List[str], required_roles: List[str]) -> bool:
    """Check if user has any of the required roles"""
    return any(role in user_roles for role in required_roles)


def has_all_roles(user_roles: List[str], required_roles: List[str]) -> bool:
    """Check if user has all of the required roles"""
    return all(role in user_roles for role in required_roles)


def is_admin(user_roles: List[str]) -> bool:
    """Check if user is an admin"""
    return ADMIN in user_roles or SUPER_ADMIN in user_roles
